File: src/src_core/fmg/core/mosaic/IMosaicController.cs
using Fmg.Common.Geom;
using Fmg.Core.Types;
using Fmg.Core.Img;
using Fmg.Core.Mosaic.Cells;

namespace Fmg.Core.Mosaic {

    /// <summary> MVC: mosaic controller interface </summary>
    /// <typeparam name="TImage">platform specific view/image/picture or other display context/canvas/window/panel</typeparam>
    /// <typeparam name="TImageInner">image type of flag/mine into mosaic field</typeparam>
    /// <typeparam name="TMosaicView">mosaic view</typeparam>
    /// <typeparam name="TMosaicModel">mosaic model</typeparam>
    public interface IMosaicController<out TImage, out TImageInner, out TMosaicView, out TMosaicModel>
                    : IImageController<    TImage,                      TMosaicView,     TMosaicModel>
        where TImage : class
        where TImageInner : class
        where TMosaicView : IMosaicView<TImage, TImageInner, TMosaicModel>
        where TMosaicModel : IMosaicDrawModel<TImageInner>
    {

        /// <summary> размер поля в ячейках </summary>
        Matrisize SizeField { get; set; }

        /// <summary> узнать тип мозаики (из каких фигур состоит мозаика поля) </summary>
        EMosaic MosaicType { get; set; }

        /// <summary> количество мин </summary>
        int MinesCount { get; set; }

        /// <summary> ячейка на которой было нажато (но не обязательно что отпущено) </summary>
        BaseCell CellDown { get; set; }

        /// <summary> Подготовиться к началу игры - сбросить все ячейки </summary>
        bool GameNew();

    }

}

File: src/src_core/fmg/core/mosaic/IMosaicController.java
package fmg.core.mosaic;

import java.util.Collection;

import fmg.common.geom.Matrisize;
import fmg.core.img.IImageController;
import fmg.core.mosaic.cells.BaseCell;
import fmg.core.types.EGameStatus;
import fmg.core.types.EMosaic;

/** MVC: mosaic controller interface
 * @param <TImage> platform specific view/image/picture or other display context/canvas/window/panel
 * @param <TImageInner> image type of flag/mine into mosaic field
 * @param <TMosaicView> mosaic view
 * @param <TMosaicModel> mosaic model
 */
public interface IMosaicController<TImage, TImageInner,
                                   TMosaicView extends IMosaicView<TImage, TImageInner, TMosaicModel>,
                                   TMosaicModel extends IMosaicDrawModel<TImageInner>>
          extends IImageController<TImage, TMosaicView, TMosaicModel>
{

    /** размер поля в ячейках */
    Matrisize getSizeField();
    /** размер поля в ячейках */
    void setSizeField(Matrisize newSizeField);

    /** узнать тип мозаики
     * (из каких фигур состоит мозаика поля) */
    EMosaic getMosaicType();
    /** установить тип мозаики */
    void setMosaicType(EMosaic newMosaicType);

    /** ячейка на которой было нажато (но не обязательно что отпущено) */
    BaseCell getCellDown();
    /** ячейка на которой было нажато (но не обязательно что отпущено) */
    void setCellDown(BaseCell cellDown);

    /** количество мин */
    int getCountMines();
    /** количество мин */
    void setCountMines(int newCountMines);

    /** сколько ещё осталось открыть мин */
    int getCountMinesLeft(); // return getCountMines() - getCountFlag();

    int getCountOpen();
    int getCountFlag();
    int getCountUnknown();


    /** Подготовиться к началу игры - сбросить все ячейки */
    boolean gameNew();
    /** Начать игру, т.к. произошёл первый клик на поле */
    void gameBegin(BaseCell firstClickCell);
    /** Завершить игру */
    Collection<BaseCell> gameEnd(boolean victory);

    /**
     *<br> Этапы игры:
     *<br>           GameNew()      GameBegin()     GameEnd()      GameNew()
     *<br>    time      |               |               |             |
     *<br>  -------->   | eGSCreateGame |               |             |
     *<br>              |  or eGSReady  |    eGSPlay    |   eGSEnd    |
     *<br>              \------ 1 -----/ \----- 2 -----/ \---- 3 ----/
     *<br>
     *<br> @see fmg.core.types.EGameStatus
     *<br>
     *<br> PS: При этапе gsReady поле чисто - мин нет! Мины расставляются только после первого клика
     *<br>     Так сделал только лишь потому, чтобы первый клик выполнялся не на мине. Естественно
     *<br>     это не относится к случаю, когда игра была создана пользователем или считана из файла.
     */
    EGameStatus getGameStatus();

}
